sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java

```diff
@@ -5,114 +5,162 @@
 
-        timer.schedule(new ReceiveTimeoutTask(work.getId(), this::dispose), work.getTimeout().toMillis());
+        // This will trigger subscription.request(N) and queue up the work
+        queueWork(initialWork);
     }
 
     /**
      * Publishes the event to the current {@link SynchronousReceiveWork}. If that work item is complete, will dispose of
      * the subscriber.
-     *
-     * @param value Event to publish.
+     * @param message Event to publish.
      */
     @Override
-    protected void hookOnNext(ServiceBusReceivedMessageContext value) {
-        work.next(value);
+    protected void hookOnNext(ServiceBusReceivedMessageContext message) {
+        currentWork.next(message);
 
-        if (work.isTerminal()) {
-            logger.info("[{}] Completed. Closing Flux and cancelling subscription.", work.getId());
-            dispose();
+        if (currentWork.isTerminal()) {
+            logger.info("[{}] Completed. Closing Flux and cancelling subscription.", currentWork.getId());
+            completeCurrentWork(currentWork);
         }
     }
 
-    @Override
-    protected void hookOnComplete() {
-        logger.info("[{}] Completed. No events to listen to.", work.getId());
-        dispose();
+    private void completeCurrentWork(SynchronousReceiveWork currentWork) {
+
+        if (isTerminated()) {
+            return;
+        }
+
+        currentWork.complete();
+        logger.verbose("[{}] work completed.", currentWork.getId());
+
+        if (timeoutOperation != null && !timeoutOperation.isDisposed()) {
+            timeoutOperation.dispose();
+        }
+        if (drainQueueDisposable != null && !drainQueueDisposable.isDisposed()) {
+            drainQueueDisposable.dispose();
+        }
+
+        if (wip.decrementAndGet() != 0) {
+            logger.warning("There is another worker in drainLoop. But there should only be 1 worker.");
+        }
+
+        // After current work finished and there more receive requests
+        if (workQueue.size() > 0) {
+            drain();
+        }
+    }
+
+    void queueWork(SynchronousReceiveWork work) {
+
+        logger.info("[{}] Pending: {}, Scheduling receive timeout task '{}'.", work.getId(), work.getNumberOfEvents(),
+            work.getTimeout());
+        workQueue.add(work);
+        drain();
+    }
+
+    private void drain() {
+        // If someone is already in this loop, then we are already clearing the queue.
+        if (!wip.compareAndSet(0, 1)) {
+            return;
+        }
+        // Drain queue..
+        drainQueueDisposable = Mono.just(true)
+            .subscribe(l -> {
+                drainQueue();
+            });
+    }
+
+    private void drainQueue() {
+        if (isTerminated()) {
+            return;
+        }
+        currentWork = workQueue.poll();
+        if (currentWork == null) {
+            return;
+        }
+
+        subscription.request(currentWork.getNumberOfEvents());
```

Comment:
What if we requested 100 events, only 30 showed up in the timeout period, and the next work item wants 10? This logic will add another 10 credits (making it 80 credits) even though we don't need to add more credits to the link.